inference-engine/src/mkldnn_plugin/nodes/mkldnn_matrix_nms_node.cpp

```diff
@@ -0,0 +1,512 @@
+        auto idx_a = candidate_index[actual_index];
+        for (int64_t j = 0; j < actual_index; j++) {
+            auto idx_b = candidate_index[j];
+            auto iou = intersectionOverUnion<T>(boxes_data + idx_a * box_size,
+                                                boxes_data + idx_b * box_size,
+                                                normalized);
+            max_iou = std::max(max_iou, iou);
+            iou_matrix[actual_index * (actual_index - 1) / 2 + j] = iou;
+        }
+        iou_max[actual_index] = max_iou;
+    });
+
+    if (scores_data[candidate_index[0]] > post_threshold) {
+        auto box_index = candidate_index[0];
+        auto box = boxes_data + box_index * box_size;
+        filtBoxes[0].box.x1 = box[0];
+        filtBoxes[0].box.y1 = box[1];
+        filtBoxes[0].box.x2 = box[2];
+        filtBoxes[0].box.y2 = box[3];
+        filtBoxes[0].index = batch_idx * boxes_num + box_index;
+        filtBoxes[0].score = scores_data[candidate_index[0]];
+        filtBoxes[0].batch_index = batch_idx;
+        filtBoxes[0].class_index = class_idx;
+        num_det++;
+    }
+
+    decay_score<T, gaussian> decay_fn;
+    for (int64_t i = 1; i < original_size; i++) {
+        T min_decay = 1.;
+        for (int64_t j = 0; j < i; j++) {
+            auto max_iou = iou_max[j];
+            auto iou = iou_matrix[i * (i - 1) / 2 + j];
+            auto decay = decay_fn(iou, max_iou, sigma);
+            min_decay = std::min(min_decay, decay);
+        }
+        auto ds = min_decay * scores_data[candidate_index[i]];
+        if (ds <= post_threshold)
+            continue;
+        auto box_index = candidate_index[i];
+        auto box = boxes_data + box_index * box_size;
+        filtBoxes[num_det].box.x1 = box[0];
+        filtBoxes[num_det].box.y1 = box[1];
+        filtBoxes[num_det].box.x2 = box[2];
+        filtBoxes[num_det].box.y2 = box[3];
+        filtBoxes[num_det].index = batch_idx * boxes_num + box_index;
+        filtBoxes[num_det].score = ds;
+        filtBoxes[num_det].batch_index = batch_idx;
+        filtBoxes[num_det].class_index = class_idx;
+        num_det++;
+    }
+    return num_det;
+}
+}//  Anonymous matrix_nms namespace
+
+void MKLDNNMatrixNmsNode::execute(mkldnn::stream strm) {
+    const float *boxes = reinterpret_cast<const float *>(getParentEdgeAt(NMS_BOXES)->getMemoryPtr()->GetPtr());
+    const float *scores = reinterpret_cast<const float *>(getParentEdgeAt(NMS_SCORES)->getMemoryPtr()->GetPtr());
+
+    const int box_shape = 4;
+    size_t  real_num_classes = m_background_class == -1 ? m_num_classes : m_num_classes - 1;
+    size_t  real_num_boxes = m_nms_top_k == -1 ? m_num_boxes : std::min(m_nms_top_k, static_cast<int>(m_num_boxes));
+    std::vector<int64_t> num_per_batch(m_num_batches);
+    std::vector<BoxInfo> filtered_boxes(m_num_batches * real_num_classes * real_num_boxes);
+
+    InferenceEngine::parallel_for(m_num_batches, [&](size_t batch) {
```

Comment:
maybe ```parallel_for2d```?

Reply:
We use parallel_for here, because he matrix_nms needs to sort the boxes among all classes inside each batch.